cdc/sink/flowcontrol/flow_control.go

```diff
@@ -110,6 +119,59 @@ func (c *TableFlowController) Release(resolvedTs uint64) {
 	c.memoryQuota.release(nBytesToRelease)
 }
 
+func (c *TableFlowController) enqueueSingleMsg(msg *model.PolymorphicEvent, size uint64) {
+	commitTs := msg.CRTs
+	lastCommitTs := atomic.LoadUint64(&c.lastCommitTs)
+
+	c.queueMu.Lock()
+	defer c.queueMu.Unlock()
+
+	var e deque.Elem
+	// 1. Processing a new transaction.
+	if e = c.queueMu.queue.Back(); e == nil || lastCommitTs < commitTs {
+		c.queueMu.queue.PushBack(&txnSizeEntry{
+			startTs:  msg.StartTs,
+			commitTs: commitTs,
+			size:     size,
+			rowCount: 1,
+		})
+		c.txnsWithSameCommitTs = 1
+		atomic.StoreUint64(&c.lastCommitTs, commitTs)
+		return
+	}
+
+	// Processing txns with the same commitTs.
+	txnEntry := e.(*txnSizeEntry)
+	if txnEntry.commitTs != lastCommitTs {
+		log.Panic("got wrong commitTs from deque, report a bug",
+			zap.Uint64("lastCommitTs", c.lastCommitTs),
+			zap.Uint64("commitTsInDeque", txnEntry.commitTs))
+	}
+
+	// 2. Append row to current transaction entry.
+	if txnEntry.startTs == msg.Row.StartTs &&
+		txnEntry.rowCount < maxRowsPerTxn && txnEntry.size < maxSizePerTxn {
+		txnEntry.size += size
+		txnEntry.rowCount++
+		return
+	}
+
+	// 3. Split the txn or handle a new txn with the same commitTs.
+	c.queueMu.queue.PushBack(&txnSizeEntry{
+		startTs:  msg.StartTs,
+		commitTs: commitTs,
+		size:     size,
+		rowCount: 1,
+	})
+	c.txnsWithSameCommitTs++
+	// mark the first data of new txnSizeEntry
+	msg.Row.SplitTxn = true
+	if c.txnsWithSameCommitTs >= batchSize {
+		// TODO(CharlesCheung): add batch resolve mechanism to mitigate oom problem
+		log.Debug("emit batch resolve event throw callback")
+	}
```

Comment:
`c.txnsWithSameCommitTs` is a little confusing, it could be 
- either different transactions with same commit-ts but different start-ts
- or the same transactions but exceeds max rows or max size, which is the logic-2 in L151

Reply:
Therefore, the name is based on the common characteristics of both cases having the same commitTs. Perhaps we should add comments to illustrate this. BTW, do you have any good names to recommend.

Reply:
What about `batchGroupCount`, and add comment to explain that batch group means a group of batch rows, all row changes have the same commit-ts.